PerfCi Grafana Dashboard: add grafonnet support

ebattat opened the pull request:
## Type of change
Note: Fill **x** in []
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] dependencies

## Description
<!--- Describe your changes below -->
Adding grafonnet support:

1. Dockerfile: using **terraform** for generating dashboard.json from grafonnet (main.libsonnet)
2. Update Grafana dashboard using generated dashboard.json 
3. Updating automatically new product versions from ElasticSearch and update main.libsonnet
4. Update Build and Nightly workflows with update_grafana_dashboard job that update Grafana dashboard
5. Add test for fetching all dashboards from Grafana for testing connectivity.
6. jsonnet/jsonnetfile.lock.json: This file configure specific grafonnet version and not with latest, that will reduce grafonnet generator failures.

[ **Important**: Currently the implementation is against testing dashboard, once all working update PerfCi dashboard ]

## For security reasons, all pull requests need to be approved first before running any automated CI

RobertKrawitz commented on benchmark_runner/common/grafana/update_grafana_last_value_mappings.py:
```diff
@@ -0,0 +1,117 @@
+        mapping_lines = []
+
+        # Read the content of the file and store lines between start and end patterns in a list
+        with open(self.main_libsonnet_path, 'r') as file:
```
Ugh.  You're editing Python code this way?  If that's what you're doing, I suggest using a clearly named comment that won't be duplicated at the end of the line, something like
```
# START_92734987
```
or some other fixed cookie.

ebattat replied:
Good point, fix it by adding
```
        //START_VALUE_MAPPING_227
        //END_VALUE_MAPPING_227

```

RobertKrawitz commented on benchmark_runner/common/grafana/update_grafana_last_value_mappings.py:
```diff
@@ -0,0 +1,117 @@
+        :return:
+        """
+        # Pattern to check in each line
+        start = "        stateTimeline.valueMapping.RegexMap.withOptions("
+        end = "+ stateTimeline.valueMapping.RegexMap.withType('value')"
```
Again, find a more robust way to do this.

ebattat replied:
Fixed, see above

RobertKrawitz commented on benchmark_runner/common/grafana/update_grafana_last_value_mappings.py:
```diff
@@ -0,0 +1,117 @@
+
+        # Replace the lines between start and end with the dictionary
+        if start_index is not None and end_index is not None:
+            lines[start_index + 1:end_index] = ['\t\t\t\t' + json.dumps(self.value_mappings, indent=None) + '\n \t\t) \n']
```
Why do you need the tabs?  And if you do, you do not need a space after the newline immediately before the tab.

ebattat replied:
need to indent the value mapping

RobertKrawitz commented on benchmark_runner/grafana/perf/Makefile:
```diff
@@ -0,0 +1,6 @@
+	@cd jsonnet/ && \
+		jb install && \
+		cd ../ && \
+		terraform apply -auto-approve
```
Need a newline at the end.

ebattat replied:
Fixed

RobertKrawitz commented on benchmark_runner/grafana/perf/dashboard.tf:
```diff
@@ -0,0 +1,20 @@
+
+output "dashboard" {
+  value = data.jsonnet_file.dashboard.rendered
+}
```
Add trailing newline.

ebattat replied:
Fixed

RobertKrawitz commented on benchmark_runner/grafana/perf/extract_dashboard.py:
```diff
@@ -0,0 +1,10 @@
+import re
+import argparse
+
+parser = argparse.ArgumentParser(description="Extract text between <<EOT and EOT in input file.")
```
Is this supposed to extract it even if not a line of its own (in the latter case) and at the end of the line (in the former)?

ebattat replied:
The terrafom will be failed if no line

RobertKrawitz replied:
I mean if EOT is somehow present inside a line as opposed to being part of a here document.

ebattat replied:
That generates by Terraform it also contains a strange prefix, this script do the work

RobertKrawitz commented on benchmark_runner/grafana/perf/jsonnet/g.libsonnet:
```diff
@@ -0,0 +1 @@
+import 'github.com/grafana/grafonnet/gen/grafonnet-latest/main.libsonnet'
```
Is there something special about g.libsonnet that you can't just have something else use main.gsonnet?

ebattat replied:
All is according to [grafonnet](https://github.com/grafana/grafonnet/blob/main/examples/terraform/jsonnet/g.libsonnet)

RobertKrawitz commented on benchmark_runner/grafana/perf/jsonnet/jsonnetfile.json:
```diff
@@ -0,0 +1,15 @@
+    }
+  ],
+  "legacyImports": true
+}
```
Add trailing newline

ebattat replied:
Fixed

RobertKrawitz commented on benchmark_runner/grafana/perf/jsonnet/main.libsonnet:
```diff
@@ -0,0 +1,7045 @@
+local g = import 'g.libsonnet';
```
Not reviewing this file.

ebattat replied:
Yes, its the grafonnet code

RobertKrawitz commented on benchmark_runner/grafana/update_grafana_dashboard.py:
```diff
@@ -0,0 +1,24 @@
+grafana_operations.increment_dashboard_version()
+grafana_operations.override_dashboard()
+
+# Error: 412
+# The 412 status code is used when a newer dashboard already exists (newer, its version is greater than the version that was sent). The same status code is also used if another dashboard exists with the same title.
```
What does this comment refer to (this looks like the entire file)?

ebattat replied:
If We got this error, to explain the reason

RobertKrawitz replied:
It's just a comment at the very end of a file.  This comment should go with whatever can actually produce that error, or better yet, if the error is received, there should be an error message.

ebattat replied:
You right, Move the error code message it to grafana_operation.py where it raised

RobertKrawitz commented on tests/integration/benchmark_runner/common/grafana/test_grafana_operations.py:
```diff
@@ -0,0 +1,18 @@
+        all_dashboards = grafana_operations.fetch_all_dashboards()
+        assert all_dashboards
+    else:
+        raise Exception('incorrect grafana url')
```
Add trailing newline.

ebattat replied:
Fixed

RobertKrawitz reviewed (commented):
You may want Thiru and/or Kelly to review this too.

RobertKrawitz reviewed (changes requested):
Needs a number of changes as specified.

ebattat commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+          git checkout main
+          git config --global user.email ${{ secrets.EMAIL }}
+          git config --global user.name  ${{ secrets.USER_NAME }}
+          git add $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard.json
```
Fixed

ebattat commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+      - name: Commit dashboard.json if any changes are detected
+        if: steps.check_changes.outputs.changes == 'true'
+        run: |
+          echo ${{ steps.check_changes.outputs.changes }}
```
Yes, Removed

ebattat commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+        run: |
+          python $GITHUB_WORKSPACE/benchmark_runner/grafana/update_grafana_dashboard.py
+          # revert dashboard.json after python version change
+          cp $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard_backup.json $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard.json
```
Fixed

ebattat commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+        id: check_changes
+        run: |
+          # backup dashboard.json
+          cp $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard.json $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard_backup.json
```
Fixed

ebattat replied:
Update the reason of file backup in comment

ebattat commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+
+      - name: Generate grafana dashboard.json using grafonnet container
+        run: |
+          sudo docker run --rm --name run_grafonnet -v $GITHUB_WORKSPACE/benchmark_runner/grafana/perf:/app --privileged quay.io/ebattat/run_grafonnet:latest
```
Fixed

RobertKrawitz commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,57 @@ jobs:
+        id: check_changes
+        run: |
+          # Backup dashboard.json before updating the Grafana dashboard to avoid changes in file indentation
+          cp -p "$GITHUB_WORKSPACE"/benchmark_runner/grafana/perf/dashboard.json "$GITHUB_WORKSPACE"/benchmark_runner/grafana/perf/dashboard_backup.json
```
Can't you put this line in the update Grafana dashboard immediately before the `python "$GITHUB...` line, before the file could be modified?

ebattat replied:
Fixed it

RobertKrawitz commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,57 @@ jobs:
+      - name: Check for changes in dashboard.json
+        id: check_changes
+        run: |
+          git diff --quiet "$GITHUB_WORKSPACE"/benchmark_runner/grafana/perf/dashboard.json || echo "changes=true" >> $GITHUB_OUTPUT
```
"$GITHUB_OUTPUT" also needs to be quoted.

RobertKrawitz commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -217,11 +217,58 @@ jobs:
+          git add $GITHUB_WORKSPACE/benchmark_runner/grafana/perf/dashboard.json
+          git commit -m "Update grafana json file"
+
+          # Pull the latest changes from the remote main branch
+          git pull origin main
```
Comment still applies.

RobertKrawitz commented on .github/workflows/Perf_Env_Build_Test_CI.yml:
```diff
@@ -261,9 +264,54 @@ jobs:
+      - name: Check for changes in dashboard.json
+        id: check_changes
+        run: |
+          git diff --quiet "$GITHUB_WORKSPACE"/benchmark_runner/grafana/perf/dashboard.json || echo "changes=true" >> $GITHUB_OUTPUT
```
"$GITHUB_OUTPUT"

RobertKrawitz commented on benchmark_runner/common/grafana/grafana_operations.py:
```diff
@@ -0,0 +1,102 @@
+
+class GrafanaOperations:
+    """
+    This class responsible for Grafana operations
```
Comment still applies.

RobertKrawitz commented on benchmark_runner/common/grafana/grafana_operations.py:
```diff
@@ -0,0 +1,102 @@
+        :return:
+        """
+        dashboard_list = []
+        headers = {
```
Comment still applies.

RobertKrawitz replied:
Shouldn't the http request have a Content-Type: header since you expect JSON?

ebattat replied:
The Content-Type header is typically used when sending data in a request (e.g., in a POST or PUT request) to specify the format of the data being sent. Since I am only fetching data using a GET request and not sending any data, I don't need to include the Content-Type header in this case.
You can see that I use content type json={"dashboard": self.dashboard_data} in override_dashboard() below

RobertKrawitz commented on benchmark_runner/grafana/perf/jsonnet/g.libsonnet:
```diff
@@ -0,0 +1,2 @@
+// Cannot import it in main.libsonnet; it must be in a separate file, g.libsonnet for more details: https://github.com/grafana/grafonnet/tree/main/examples/terraform
```
That's just a pointer to an example.  Note that attempting to import it inside main.libsonnet results in an error.

ebattat replied:
Updated

RobertKrawitz reviewed (changes requested):
Still have some comments to address.

RobertKrawitz commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -314,3 +315,49 @@ jobs:
+        if: steps.check_changes.outputs.changes == 'true'
+        run: |
+          git checkout main
+          git config --global user.email ${{ secrets.EMAIL }}
+          git config --global user.name  ${{ secrets.USER_NAME }}
```
Should be

```
git config --global user.email "${{ secrets.EMAIL }}"
```

ebattat replied:
Fixed

RobertKrawitz commented on benchmark_runner/common/grafana/update_grafana_last_value_mappings.py:
```diff
@@ -0,0 +1,117 @@
+            data = self.elasticsearch.get_elasticsearch_index_by_id(index='ci-status', id=id)
+            for version, value in data['_source'].items():
+                if version in display_versions and value not in new_versions.values():
+                    new_versions[value.replace(".", "").replace("-", "")] = value
```
Not a big deal; you can do regexp stuff in python too.

RobertKrawitz reviewed (changes requested):
Just a few more quoting changes and one Content-Type:

ebattat commented:
@RobertKrawitz, The parser works fine and parse it without EOT

RobertKrawitz commented:
> @RobertKrawitz, The parser works fine and parse it without EOT

Just because it works in a test doesn't mean that it's robust.  You need to handle edge cases and unexpected input too.

ebattat commented:
@RobertKrawitz here is an example for the code we parse

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

```
Outputs:

dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   "graphTooltip": 1,
   "panels": [
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "reqps"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 24
         },
         "id": 1,
         "targets": [
            {
               "datasource": {
                  "type": "prometheus",
                  "uid": "ops-cortex"
               },
               "expr": "sum by (status_code) (rate(request_duration_seconds_count{job=~\".*/faro-api\"}[$__rate_interval]))"
            }
         ],
         "title": "Requests / sec",
         "type": "timeseries"
      }
   ],
   "schemaVersion": 36,
   "time": {
      "from": "now-6h",
      "to": "now"
   },
   "timezone": "utc",
   "title": "Faro dashboard"
}

EOT
```

RobertKrawitz commented:
What happens if the original is:

```
{
   "description": "Dashboard for Faro",
   "graphTooltip": 1,
   "panels": [
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "reqps"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 24
         },
         "id": 1,
         "targets": [
            {
               "datasource": {
                  "type": "prometheus",
                  "uid": "EOT-cortex"
               },
               "expr": "sum by (status_code) (rate(request_duration_seconds_count{job=~\".*/faro-api\"}[$__rate_interval]))"
            }
         ],
         "title": "Requests / sec",
         "type": "timeseries"
      }
   ],
   "schemaVersion": 36,
   "time": {
      "from": "now-6h",
      "to": "now"
   },
   "timezone": "utc",
   "title": "Faro dashboard"
}
```

You'll then get
```
{
   "description": "Dashboard for Faro",
   "graphTooltip": 1,
   "panels": [
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "reqps"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 24
         },
         "id": 1,
         "targets": [
            {
               "datasource": {
                  "type": "prometheus",
                  "uid": "
```

as your output.

I know this is unlikely, but the code needs to be written defensively.

ebattat commented:
@RobertKrawitz. so what do you recommend ?

RobertKrawitz commented on .github/workflows/Nightly_Perf_Env_CI.yml:
```diff
@@ -314,3 +315,84 @@ jobs:
+          ELASTICSEARCH_USER: ${{ secrets.PERF_ELASTICSEARCH_USER }}
+          ELASTICSEARCH_PASSWORD: ${{ secrets.PERF_ELASTICSEARCH_PASSWORD }}
+        run: |
+          python $GITHUB_WORKSPACE/grafonnet_generator/grafana/update_versions_main_libsonnet.py
```
Needs to be quoted.

ebattat replied:
Fixed

RobertKrawitz commented on .github/workflows/Perf_Env_Build_Test_CI.yml:
```diff
@@ -261,9 +264,89 @@ jobs:
+          ELASTICSEARCH_USER: ${{ secrets.PERF_ELASTICSEARCH_USER }}
+          ELASTICSEARCH_PASSWORD: ${{ secrets.PERF_ELASTICSEARCH_PASSWORD }}
+        run: |
+          python $GITHUB_WORKSPACE/grafonnet_generator/grafana/update_versions_main_libsonnet.py
```
Needs to be quoted.

ebattat replied:
Fixed

RobertKrawitz reviewed (changes requested):
Few more quotes, and still need to fix the EOT regex.

ebattat commented:
@RobertKrawitz 

1. Http content type:
The Content-Type header is typically used when sending data in a request (e.g., in a POST or PUT request) to specify the format of the data being sent. Since I am only fetching data using a GET request and not sending any data, I don't need to include the Content-Type header in this case.
You can see that I use content type json={"dashboard": self.dashboard_data} in override_dashboard() method below

2. EOT
I have added dictionary validation by load it to json, meaning if we have the following example, we will get the following error: Extracted content is not valid JSON.

```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   "graphTooltip": "EOT",
   "title": "Faro dashboard"
}

EOT
"""
```

RobertKrawitz commented:
> @RobertKrawitz
> 
> 1. Http content type:
>    The Content-Type header is typically used when sending data in a request (e.g., in a POST or PUT request) to specify the format of the data being sent. Since I am only fetching data using a GET request and not sending any data, I don't need to include the Content-Type header in this case.

OK.

> 2. EOT
>    I have added dictionary validation by load it to json, meaning if we have the following example, we will get the following error: Extracted content is not valid JSON.

What you should do is improve the regexp so that it correctly handles this situation, not raise an error because a string happened to be within the text.

ebattat commented:
@RobertKrawitz.

The only way to handle it is to inject a special string instead of EOT, e.g. `@@EOT@@`.
We are using existing terraform jsonnet parser and we cannot change it[
](https://registry.terraform.io/providers/alxrem/jsonnet/latest/docs)
So the current solution is the best we can do, if there will be any parsing issue, exception will be raised.

RobertKrawitz commented:
> @RobertKrawitz.
> 
> The only way to handle it is to inject a special string instead of EOT, e.g. `@@EOT@@`. We are using existing [terraform jsonnet parser code ](https://registry.terraform.io/providers/alxrem/jsonnet/latest/docs) that generate auto the following output syntax:
> 
> ```
> dashboard = <<EOT
> {
>    "description": "Dashboard for Faro",
>    "graphTooltip": "aaa",
>    "title": "Faro dashboard"
> }
> 
> EOT
> ```
> 
> So the current solution is the best we can do, if there will be any parsing issue, exception will be raised.

No, the regular expression can be more intelligent, to only recognize the second EOT on a line by itself.  Rather than

```
r'dashboard = <<EOT(.*?)EOT'
```

it can be something like

```
r`dashboard = <<EOT(.*\n)EOT($|\n)'
```
to pick up the final EOT on a line by itself, either at the very end or between two newlines.

RobertKrawitz commented:
The other way to do it is to pass it to a shell to parse, but then you have to trust that the code isn't going to do something it shouldn't.  Parsing the here doc is reasonable enough, just do it correctly.

RobertKrawitz commented on benchmark_runner/grafana/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+ENV PATH="/root/go/bin:${PATH}"
+
+CMD ["/bin/sh", "-c", "terraform init; make > dashboard.txt;  python3 extract_dashboard.py dashboard.txt dashboard.json; terraform destroy -auto-approve > /dev/null 2>&1"]
```
I think there's a much easier way you can do this:

```
CMD ["/bin/sh", "-c", "terraform init; make |/bin/sh > dashboard.json; terraform destroy -auto-approve > /dev/null 2>&1"]
```
and then you won't need extract_dashboard.py at all.

ebattat replied:
Tried, but its not working:

```
[root@6e8e49594c69 app]# make |/bin/sh > dashboard.json
/bin/sh: line 1: $'\E[0m\E[1mdata.jsonnet_file.dashboard:': command not found
```

ebattat replied:
I have updated the extract code for suffix: **\nEOT\n**
`r'dashboard = <<EOT(.*?)\nEOT\n'`
That should solve the EOT issue.

RobertKrawitz replied:
I think you need to run `terraform init --no-color`.

RobertKrawitz replied:
From https://developer.hashicorp.com/terraform/cli/commands/init
```
-no-color Disable color codes in the command output.
```

You're much better off not trying to fight what terraform init is trying to do here.

RobertKrawitz replied:
What exactly does `make` produce here?  Spaces between the variable name, the `=`, and the text are not valid shell, so maybe it's not producing what I think it is and that regex may be the only way to go.  But I'd like to see both the `Makefile` and the output.

ebattat replied:
The `terraform init` is not related to the problem at all.
The problem is with the `terraform apply` inside Makefile.
This command converts jsonnet(grafonnet) code into json and its using [Jsonnet terraform rendered code](https://registry.terraform.io/providers/alxrem/jsonnet/latest/docs) that rendered it into constant output
```
dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   "graphTooltip": "aaa",
   "title": "Faro dashboard"
}

EOT
```

ebattat replied:
```
.PHONY: apply
apply:
        @cd jsonnet/ && \
                jb install && \
                cd ../ && \
                terraform apply -auto-approve > dashboard.txt && \
                python3 extract_dashboard.py dashboard.txt dashboard.json
```
@RobertKrawitz, I put all inside Makefile, WDYT ?

RobertKrawitz replied:
Might as well put it all in the makefile; I still think you should avoid that temporary file (dashboard.txt).

RobertKrawitz replied:
Where does `/root/grafonnet` come from?

ebattat replied:
terraform apply -auto-approve -json => generates corrupted json output

ebattat replied:
/root/grafonnet/ => You should put your local path.

RobertKrawitz replied:
Local path to what?

Can you include the generated JSON?

ebattat replied:
[dashboard.json](https://github.com/redhat-performance/benchmark-runner/blob/3ec88b4167cca890904c6211866d51277f1675fb/benchmark_runner/grafana/perf/dashboard.json)
I have pushed the changes, think its good enough for now

RobertKrawitz commented on benchmark_runner/grafana/perf/extract_dashboard.py:
```diff
@@ -0,0 +1,19 @@
+import re, sys, json
+
+if len(sys.argv) != 3:
+    raise ValueError('Usage: python extract_dictionary.py input_file output_file')
```
It's called extract_dashboard.  But you really should have it read from stdin and write to stdout, without any temporary files.

ebattat replied:
You right, remove temporary file, read directly from stdin

`terraform apply -auto-approve | python3 extract_dashboard.py dashboard.json`

RobertKrawitz commented on benchmark_runner/grafana/perf/Makefile:
```diff
@@ -0,0 +1,5 @@
+	@cd jsonnet/ && \
+		jb install && \
+		cd ../ && \
+		terraform apply -auto-approve | python3 extract_dashboard.py dashboard.json
```
It would be best for extract_dashboard.py to simply send its output to stdout (so you could easily e. g. pipe it through `jq` if desired), but I'll accept this, as it avoids the temporary file.

RobertKrawitz replied:
Yes, I approved it as such.  I still think that extract_dashboard should write to stdout rather than to a separate file; that would allow you to e. g. process it further through jq, but I've accepted it as it avoids the temporary file.

ebattat replied:
Do you mean to dashboard.json file?
We must have this file because we use it to update Grafana dashboard in case of changed.

RobertKrawitz replied:
Correct, but you could write it as

```
terraform apply -auto-approve | python3 extract_dashboard.py > dashboard.json
```

Then if you want at some point to process the JSON with jq downstream of extract_dashboard.py you can simply add it to the pipeline.

ebattat replied:
You right, updated it !!!

RobertKrawitz commented on benchmark_runner/common/grafana/update_grafana_last_value_mappings.py:
```diff
@@ -0,0 +1,116 @@
+    """
+    This class updates grafana dashboard with last value mappings from ElasticSearch
+    """
+    MAX_VERSION_LEN = 20
```
What's this for (please document it).

ebattat replied:
@RobertKrawitz, add document.
Prevent error messages from being displayed instead of versions in case of a connection issue

ebattat commented:
@RobertKrawitz,

I have added jsonnetfile.lock.json file for working with specific grafonnet and not the latest grafonnet version.
That should reduce failure while using grafonnet generator. 
I have added it due to recent grafonnet code [fix](https://github.com/grafana/grafonnet/commit/2359881e8a9e5601e3ccf8f7f2b024f06a70adf0) that cause a failure when generating dashboard.json. 
WDYT ?

RobertKrawitz commented on benchmark_runner/grafana/perf/jsonnet/jsonnetfile.lock.json:
```diff
@@ -0,0 +1,86 @@
+{
```
Rather than `.lock`, I suggest `.version` if that's possible (i. e. if jsonnet doesn't have some kind of hard-coded requirement for that filename).  `lock` implies that it's locking against some kind of concurrency issue; this is about versioning, not concurrency.

RobertKrawitz replied:
That looks fine.

RobertKrawitz reviewed (changes requested):
Unless this is required by jsonnet, please change the version file name from jsonnetfile.lock.json to jsonnetfile.version.json or similar.  `lock` implies concurrency control of some kind, which isn't what this is about.
